api/v1alpha1/healthcheck_types.go

```diff
@@ -0,0 +1,149 @@
+	//
+	// +kubebuilder:validation:Format=duration
+	Interval metav1.Duration `json:"interval"`
+
+	// UnhealthyThreshold defines the number of unhealthy health checks required before a backend host is marked unhealthy.
+	//
+	// +kubebuilder:validation:Minimum=1
+	UnhealthyThreshold uint32 `json:"unhealthyThreshold"`
+
+	// HealthyThreshold defines the number of healthy health checks required before a backend host is marked healthy.
+	//
+	// +kubebuilder:validation:Minimum=1
+	HealthyThreshold uint32 `json:"healthyThreshold"`
+
+	// HealthChecker defines the concrete health checker to do health checking.
+	HealthChecker HealthChecker `json:"healthChecker"`
+}
+
+// HealthCheckerType is the type of health checker.
+// +kubebuilder:validation:Enum=HTTP;GRPC;TCP
+type HealthCheckerType string
+
+const (
+	// HealthCheckerTypeHTTP defines the HTTP type of health checking.
+	HealthCheckerTypeHTTP HealthCheckerType = "HTTP"
+	// HealthCheckerTypeGRPC defines the GRPC type of health checking.
+	HealthCheckerTypeGRPC HealthCheckerType = "GRPC"
+	// HealthCheckerTypeTCP defines the TCP type of health checking.
+	HealthCheckerTypeTCP HealthCheckerType = "TCP"
+)
+
+// HealthChecker defines the configuration of concrete health checker.
+// EG supports various types of health checking including HTTP, GRPC, TCP.
+// +union
+//
+// +kubebuilder:validation:XValidation:rule="self.type == 'HTTP' ? has(self.http) : !has(self.http)",message="If Health Checker type is HTTP, http field needs to be set."
+// +kubebuilder:validation:XValidation:rule="self.type == 'GRPC' ? has(self.grpc) : !has(self.grpc)",message="If Health Checker type is GRPC, grpc field needs to be set."
+// +kubebuilder:validation:XValidation:rule="self.type == 'TCP' ? has(self.tcp) : !has(self.tcp)",message="If Health Checker type is TCP, tcp field needs to be set."
+type HealthChecker struct {
+	// Type defines the type of health checker.
+	// +kubebuilder:validation:Enum=HTTP;GRPC;TCP
+	// +unionDiscriminator
+	Type HealthCheckerType `json:"type" yaml:"type"`
+	// HTTP defines the configuration of http health checker.
+	// It's required while the health checker type is HTTP.
+	// +optional
+	HTTP *HTTPHealthChecker `json:"http,omitempty" yaml:"http,omitempty"`
+	// GRPC defines the configuration of grpc health checker.
+	// It's required while the health checker type is GRPC.
+	// +optional
+	GRPC *GRPCHealthChecker `json:"grpc,omitempty" yaml:"grpc,omitempty"`
+	// TCP defines the configuration of tcp health checker.
+	// It's required while the health checker type is TCP.
+	// +optional
+	TCP *TCPHealthChecker `json:"tcp,omitempty" yaml:"tcp,omitempty"`
+}
+
+// HTTPHealthChecker defines the settings of http health check.
+type HTTPHealthChecker struct {
+	// Path defines the HTTP path that will be requested during health checking.
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=1024
+	Path string `json:"path" yaml:"path"`
```

Comment:
should it be `Path` or `Url` ?

Reply:
how is the user providing the destination host name ?

Reply:
> how is the user providing the destination host name ?

Envoy will use the address of each endpoint as the host name. If you are referring to the Host header, envoy offers an optional Host field to set this. Should it be supported?

Reply:
how can you reach a path without specifying the `host`